Support for memory caching in ARM MPU

aurel32 opened the pull request:
This patch series is based on the patch from #8185 and the associated discussion. It changes the MPU attributes to support cache on RAM and Flash. It should be noted that those changes are basically a no-op if the cache is not enabled.

This should address task 3 of #8784.

ioannisg commented:
@aurel32 thanks for this PR. We have a race condition with #8785. Let's decide which gets-in first.

aurel32 commented:
> We have a race condition with #8785. Let's decide which gets-in first.

Oh indeed. I don't think it's a big issue. The most difficult task is probably to agree with the changes. Rebasing one PR or the other depending on the one that gets in first should be relatively easy.

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -109,18 +109,18 @@ static inline u32_t _get_region_attr_by_type(u32_t type, u32_t size)
 	switch (type) {
 #ifdef CONFIG_USERSPACE
 	case THREAD_STACK_REGION:
-		return _get_region_attr(1, P_RW_U_RW, 0, 1, 0,
```
According to http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0553a/Cihegaib.html you are setting here "Outer and inner write-back. No write allocate". You will achieve better performance with "Outer and inner write-back. Write and read allocate"

aurel32 replied:
Indeed, it seems WBWA is used on the RAM region when the MPU is not enabled. I'll change that.

ioannisg replied:
agree with @pizi-nordic. I made a similar comment in #8185 about configuring RAM regions as Write-Back Write-Allocate (and Non-Shareable).

pizi-nordic commented on include/arch/arm/cortex_m/mpu/arm_mpu_v7m.h:
```diff
@@ -102,7 +102,7 @@
 /* Some helper defines for common regions */
 #define REGION_RAM_ATTR(size) \
-	(NORMAL_OUTER_INNER_NON_CACHEABLE_NON_SHAREABLE | \
+	(NORMAL_OUTER_INNER_WRITE_BACK_NON_SHAREABLE | \
```
Here also wtite-allocate is disabled.

pizi-nordic commented on include/arch/arm/cortex_m/mpu/arm_mpu_v7m.h:
```diff
@@ -101,19 +101,16 @@
-	MPU_RASR_XN_Msk | size | FULL_ACCESS_Msk)
 #define REGION_RAM_ATTR(size) \
-	(NORMAL_OUTER_INNER_NON_CACHEABLE_NON_SHAREABLE | \
+	(NORMAL_OUTER_INNER_WRITE_BACK_NON_SHAREABLE | \
```
Here also wtite-allocate is disabled.

aurel32 commented:
@pizi-nordic, I have just done the changes your requested.

ioannisg reviewed (approved):
Looks good to me, apart from the comment by @pizi-nordic .

aurel32 committed:
arch: arm: mpu: fix outer inner WBWA non-shareable macro

Fix NORMAL_OUTER_INNER_WRITE_BACK_WRITE_READ_ALLOCATE_NONSHAREABLE, the
bufferable bit should be set, while the shareable one should not be set.

At the same time, rename it from ..._NONSHAREABLE to _NON_SHAREABLE to
keep the same naming convention for all macro. Given it's not (yet)
used, it should not be an issue.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
arch: arm: mpu: enable WBWA caching on per thread user RAM

Update _get_region_attr_by_type used to set up the user regions
on a per thread basis to match the REGION_USER_RAM_ATTR macro. Caching
attributes were set "Outer and inner write-through. No write allocate.",
this patch changes it to "Outer and inner write-back. Write and read
allocate.".

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
arch: arm: mpu: remove REGION_USER_RAM_ATTR macro

This is currently unused, and all the user regions are set up at runtime
on a per thread basis.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
arch: arm: mpu: enable WT caching on Flash

Change the REGION_FLASH_ATTR macros to set the cache attributes on Flash
regions to "Outer and inner write-through. No write allocate.". This
matches the cache attributes used when the MPU is disabled.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
arch: arm: mpu: enable WBWA caching on RAM

Change the REGION_RAM_ATTR macro to set the cache attributes on RAM
regions to "Outer and inner write-back. Write and read allocate". This
matches the cache attributes used when the MPU is disabled.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
I have just pushed a new version, with no code change. I have just changed the subject of the second patch so that the shippable checks pass.

ioannisg commented:
Looks ok to me. @pizi-nordic , @galak?

pizi-nordic commented:
What about shareable attribute?
As far as I know Zephyr supports SMP and AMP, so in order to not break these we should keep it on.

ioannisg commented:
>What about shareable attribute?
As far as I know Zephyr supports SMP and AMP, so in order to not break these we should keep it on.

Shouldn't this be defined on a per-region basis at run-time (in case we want to share some memory with other core(s)?)

pizi-nordic commented:
>>What about shareable attribute?
>>As far as I know Zephyr supports SMP and AMP, so in order to not break these we should keep it on.
>Shouldn't this be defined on a per-region basis at run-time (in case we want to share some memory >with other core(s)?)

In AMP, only shared memory needs the shareable attribute. In SMP, all regions needs it.

ioannisg commented:
@pizi-nordic This PR is about cache-ability attribution, so I wonder whether you would agree to address the shareability configuration in a separate issue?

aurel32 commented:
> What about shareable attribute?
> As far as I know Zephyr supports SMP and AMP, so in order to not break these we should keep it on.

Note that this patch series does not turn off the shareable attribute, it was already off before.

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -109,18 +109,18 @@ static inline u32_t _get_region_attr_by_type(u32_t type, u32_t size)
 	case THREAD_APP_DATA_REGION:
-		return _get_region_attr(1, P_RW_U_RW, 0, 1, 0,
-					1, 0, region_size);
+		return _get_region_attr(1, P_RW_U_RW, 1, 1, 1,
```
@aurel32: Here shareable attribute is removed from the region.

aurel32 replied:
Oh indeed, that's a mistake, that's why it is not mentioned in the commit message. I'll fix that.

ioannisg commented:
> In AMP, only shared memory needs the shareable attribute. In SMP, all regions needs it.

Is there any ARM CPU that supports it?

pizi-nordic commented:
>>In AMP, only shared memory needs the shareable attribute. In SMP, all regions needs it.

>Is there any ARM CPU that supports it?

grep SMP arch/arm says no.

ioannisg commented:
>grep SMP arch/arm says no.

Then I suggest, we leave out the share-ability configuration from this PR

pizi-nordic commented:
Ok.

aurel32 commented:
I have just pushed a new version. The last patch is changed to not touch the share-ability. The patch only changes the cache-ability.

pizi-nordic commented on include/arch/arm/cortex_m/mpu/arm_mpu_v7m.h:
```diff
@@ -53,8 +53,8 @@
 	((1 << MPU_RASR_TEX_Pos) |\
 	 MPU_RASR_C_Msk | MPU_RASR_B_Msk | MPU_RASR_S_Msk)
-#define NORMAL_OUTER_INNER_WRITE_BACK_WRITE_READ_ALLOCATE_NONSHAREABLE \
-	((1 << MPU_RASR_TEX_Pos) | MPU_RASR_C_Msk | MPU_RASR_S_Msk)
```
Here shareable attribute is changed too.

aurel32 replied:
I disagree. First of all this change was made to the macro so that its behaviour matches its name. As noted in the commit message, this macro was unused before. Then the following patch changes the RAM attributes from `NORMAL_OUTER_INNER_NON_CACHEABLE_NON_SHAREABLE` to  `NORMAL_OUTER_INNER_WRITE_BACK_WRITE_READ_ALLOCATE_NON_SHAREABLE`.

Overall it means that the shareable attribute of the RAM region is unchanged and left as disabled.

pizi-nordic replied:
Ok

ioannisg commented:
@galak , @agross-linaro could you, please, have a look at this PR?

agross-oss reviewed (approved):
lgtm
